### PR TITLE
[HandshakeToFIRRTL] lower fork operation with FIRRTL primitives

### PR DESCRIPTION
**Pull request by hanchenye**
Previously, the ForkOp lowering is implemented with the LazyForkOp's lowering logic. In this patch, the ForkOp lowering is re-implemented using FIRRTL primitives.

### CODE REVIEW

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 1]
```diff
@@ -1356,7 +1433,8 @@ struct HandshakeFuncOpLowering : public OpConversionPattern<handshake::FuncOp> {
       // be instantiated in the top-module.
       else if (op.getDialect()->getNamespace() != "firrtl") {
         FModuleOp subModuleOp = checkSubModuleOp(topModuleOp, &op);
-        bool hasClock = isa<handshake::BufferOp, handshake::ControlMergeOp>(op);
+        bool hasClock =
```
Comment (mikeurbach) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:1436]: This list is going to keep growing with memory support. Not a requirement for this PR, but at some point maybe we could have a trait on the ops themselves to indicate if they have a clock.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp [hunk 2]
```diff
@@ -1142,37 +1142,114 @@ bool HandshakeBuilder::visitHandshake(LazyForkOp op) {
...
+
+    // Finally, we can drive the doneWire we created in the beginning with
+    // {validReadyWire || emtdReg}, where emtdReg indicates a successful
+    // handshake in a previous clock cycle.
+    rewriter.create<ConnectOp>(
+        insertLoc, doneWire,
+        rewriter.create<OrPrimOp>(insertLoc, bitType, validReadyWire, emtdReg));
+
+    // All done, move to the next result port.
+    idx++;
```
Comment (mikeurbach) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:1252]: nit: I think the LLVM style is to prefer pre-increment

**Review by mikeurbach — APPROVED** [review]
This looks good to me. Logic is clear to follow and the test cases are really nice.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp (hunk 1 above)
Comment (hanchenye) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:1459]: Good point!

**Comment by hanchenye** [comment]
Factor out a buildForkLogic() method, in this way other components (e.g., MemoryOp) who need to build fork logic can also call this method.

**Comment by mikeurbach** [comment]
Thanks! Looks good to me. I can use the new buildForkLogic when I send the MemoryOp PR.

**Comment by mikeurbach** [comment]
@hanchenye is this ready to merge? It looks good to me and I will happily use it.

**Comment by mikeurbach** [comment]
Thanks!

**Pull request merged**